### ISSUE TITLE
[$1000] When adding a secondary email, there is no timer - you can request multiple magic codes

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#24312]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to Setting,
2. Click on Avatar and go to Profile
3. Click on Contact Method
4. Click on Add New Contact Method
5. Enter valid email, click on add
6. Go back to Contact mrhtods page, and click on recently added email
7. You can see Magic code field, and an option to Resend it again
## Expected Result: 
Every time a magic code is sent, there ix a 30 second countdown timer. This does not show.
## Actual Result: 
User can click on the Request Magiv Codes link multiple times, and by 5th tr 6th time, they get a message saying they cannot send codes wait and try again.

Now, when you try to add another email, you'll get the error first time only.

The behaviour is not consistent

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.52-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/0ff4252d-1b09-45de-8d97-8c138ef6d250

**Expensify/Expensify Issue URL:**
**Issue reported by:** @premsjce
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690907049368789

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019c70432915b26b33</li>
        <li>Upwork Job ID: 1689490341495549952</li>
        <li>Last Price Increase: 2023-08-17</li>
    </ul>
</details>

**Comment by DanutGavrus** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In adding secondary email, there is no timeout function between sending magic codes.

### What is the root cause of that problem?
At login, we have implemented a timer functionality, while at adding a secondary email, we have not.

### What changes do you think we should make in order to solve the problem?
We should either implement the same logic as in `BaseValidateCodeForm` [component](https://github.com/Expensify/App/blob/main/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js):
1. Define a **timeRemaining** state;
2. Use **setTimeout** in order to update it each second;
3. Re-start the **timer** with its original value when user re-sends the code.

Or, move the timer's logic to a new *reusable component* and use it in both places.

**Comment by jliexpensify** [comment]
From my own testing, I can reproduce, but I think there's another bug with adding a secondary contact method. One additional one I've found is:

1. Remove a current contact method. You can refresh the page, if needed. Notice that it's not been added.
2. Try adding the same contact method again. Error: "Contact Method Already Exists" (even thought you removed it).
3. Try refreshing the page again and re-adding same contact method. See the same error.
4. Navigate back to the profile page and see that the contact method has been added, but only sometimes.
5. See this error:

![image](https://github.com/Expensify/App/assets/30815269/fc2bb461-232f-42ea-9091-1ee50602520c)

**Comment by jliexpensify** [comment]
@abdulrahuman5196 - I'm surfacing [this comment](https://expensify.slack.com/archives/C049HHMV9SM/p1691565944462639?thread_ts=1690907049.368789&cid=C049HHMV9SM) from the Slack thread. I don't think it's a feature request, as we've implemented a 30 second timer to prevent users requesting multiple codes at once, so I don't see why we wouldn't want this behaviour with Secondary Logins. Keen to hear if you agree!

**Comment by dotsquarespriyanka** [comment]
**Proposal**
1. **The Problem**
- Magic Code Timer is missing.
- Uncoordinated behavior after removing contact method

2. **Root Cause**
- Without a timer implemented during the process of adding secondary emails, users can repeatedly request magic codes without time limitations, which may lead to inappropriate use.

3. **Problem Solution Implementation Steps**
A- **Identify Components**
- Locate the components responsible for managing secondary email addresses and magic code requests. This might involve examining the Profile and Contact Method sections.

B. **Add Timer State**
- In the component's state, introduce a variable named timeRemaining initialized with a value (e.g., 30 seconds).

C. **Implement Timer Logic**
- Inside the component, create a function, startTimer, that uses setTimeout. This function will decrement timeRemaining every second.
const startTimer = () => {
  let remaining = timeRemaining;
  const interval = setInterval(() => {
    if (remaining > 0) {
      remaining--;
      this.setState({ timeRemaining: remaining });
    } else {
      clearInterval(interval);
    }
  }, 1000);
};

D. **Disable Link During Timer**
- Disable the "Request Magic Codes" link when the timer is active by checking the value of timeRemaining.

E. **Re-send Timer**
- Implement a similar mechanism for the "Resend" link, resetting timeRemaining to its initial value when clicked.

F. **User Feedback**
- Create a visual element (e.g., countdown timer) displaying timeRemaining to give users feedback on when they can request a new magic code.

G. **Error Handling**
- When the "Request Magic Codes" link is clicked before the timer expires, display an error message indicating that the user should wait for the timer to complete.

**Comment by dotsquarespriyanka** [comment]
@MelvinBot 
priyanka.agarwalds@gmail.com
https://www.upwork.com/freelancers/~015d4cb08826b94006

**Comment by kaushiktd** [comment]
### Please re-state the problem that we are trying to solve in this issue.

When adding a secondary email, there is no timer - you can request multiple magic codes.

### What is the root cause of that problem?

On `BaseValidateCodeForm.js` for a contact method, You need to add a timer function and offline indicator. Currently not implemented so this is the main root cause.

### What changes do you think we should make in order to solve the problem?

Replace the below-mentioned code on [here.](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L166-L177)

Need to show timer if the value is greater than 0. Otherwise, show Send Button.

```
{timeRemaining > 0 && !props.network.isOffline ? (
    <Text>
        {props.translate('validateCodeForm.requestNewCode')}
        <Text style={[styles.textBlue]}>00:{String(timeRemaining).padStart(2, '0')}</Text>
    </Text>
) : (
    <PressableWithFeedback
        disabled={props.network.isOffline}
        style={[styles.mr1]}
        onPress={resendValidateCode}
        underlayColor={themeColors.componentBG}
        hoverDimmingValue={1}
        pressDimmingValue={0.2}
        accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
        accessibilityLabel={props.translate('validateCodeForm.magicCodeNotReceived')}
    >
        <Text style={[StyleUtils.getDisabledLinkStyles(props.network.isOffline)]}>{props.translate('validateCodeForm.magicCodeNotReceived')}</Text>
    </PressableWithFeedback>
)}
```

**Show offline indicator and resend new code timer**
You need to define `timeRemaining` variable as well as `timerRef` by using useRef() So please add the below mentioned two lines at [here](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L79)

```
const [timeRemaining, setTimeRemaining] = useState(30);
const timerRef = useRef();
```

**Add new useEffect function for time**
You need to create a time counter function to calculate the remaining time. Add the below code [here.](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L80)


```
useEffect(() => {
    if (timeRemaining > 0) {
        timerRef.current = setTimeout(() => {
            setTimeRemaining(timeRemaining - 1);
        }, 1000);
    }
    return () => {
        clearTimeout(timerRef.current);
    };
}, [timeRemaining]);
```

You need to assign a value to the timer when hitting the `resendValidateCode` function after line#107 in the below file
https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L107

```
setTimeRemaining(30);
```

### Video:

https://drive.google.com/file/d/1Pdcu9bXOr-GhjXAyHv6u-c6SFq6ebH-P/view?usp=sharing

**Comment by nileshahir286** [comment]
### **Contributor details**
Your Expensify account email: [nileshahir286@gmail.com](mailto:nileshahir286@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~0126f3665fe23ff297

### **Proposal**
### **Please re-state the problem that we are trying to solve in this issue.**
Every time a magic code is sent, there ix a 30 second countdown timer. This does not show.

### **What is the root cause of that problem?**
As i have checked there is no timer added for magic code, it's on login page but not on this one.

### **What changes do you think we should make in order to solve the problem?**

We should implement the timer on  this file also BaseValidateCodeForm.js
either implement the same logic as in BaseValidateCodeForm [component].
and timer gets reset when we click on 'Did't received a magic code?' .
`       <PressableWithFeedback
                        disabled={props.network.isOffline || timeRemaining > 0}
                        style={[styles.mr1]}
                        onPress={resendValidateCode}
                        underlayColor={themeColors.componentBG}
                        hoverDimmingValue={1}
                        pressDimmingValue={0.2}
                        accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
                        accessibilityLabel={props.translate('validateCodeForm.magicCodeNotReceived')}
                    >
                            {timeRemaining > 0  ? 
                            <Text style={[styles.mt2]}>
                                {props.translate('validateCodeForm.requestNewCode')}
                                <Text style={[styles.textBlue]}>00:{String(timeRemaining).padStart(2, '0')}</Text>
                            </Text>:
                        <Text style={[StyleUtils.getDisabledLinkStyles(props.network.isOffline)]}>{props.translate('validateCodeForm.magicCodeNotReceived')}</Text> }
                    </PressableWithFeedback>`

` useEffect(() => {
        if (timeRemaining > 0) {
            timerRef.current = setTimeout(() => {
                setTimeRemaining(timeRemaining - 1);
            }, 1000);
        }
        return () => {
            clearTimeout(timerRef.current);
        };
    }, [timeRemaining]);`

`const resendValidateCode = () => {
        setTimeRemaining(30);
        User.requestContactMethodValidateCode(props.contactMethod);
        setValidateCode('');
        inputValidateCodeRef.current.focus();
    };`

**### See the video bellow**
[screen-capture.webm](https://github.com/Expensify/App/assets/6794076/704d8ca7-e84d-412b-ab44-0cd4642cc63b)

**Comment by jliexpensify** [comment]
Bump @abdulrahuman5196 we have q few proposals!

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When validating secondary contact method, the user is able to request multiple magic codes one after another and after 5th or 6th attempt, the user is blocked and a message displayed to wait few minutes before making further requests.

### What is the root cause of that problem?
The root cause here is that the current implementation allows user to request multiple magic codes one after another than the user gets blocked with a vague message to wait few minutes and try again later. It is ambigous to wait few minutes and the message can be made clear using a timer functionality like that in the signin page where a 30 second timer is shown after which the user can request another magic code.

### What changes do you think we should make in order to solve the problem?

Implement timer functionality to replace the existing functionality of requesting multiple codes one after another. After requesting magic code, the user is shown the 30 second timer to wait. The 30 second wait message is much more clearer than the previous vague message of waiting few minutes and trying again.

I am attaching the video below of functional solution. Currently the message 'link has been re-sent' is shown after requesting magic code. The existing functionality remains intact where the user is asked to wait a few minutes but if required can be removed.

https://github.com/Expensify/App/assets/141340971/7f33bda9-f608-43ca-abbe-de7bd296852c

@jliexpensify regarding the bug that you found when adding secondary contact method. I was not able to reproduce it, if you can make a video and reproduce it, I could potentially resolve it along with the timer functionality.

**Comment by abdulrahuman5196** [comment]
>I'm surfacing [this comment](https://expensify.slack.com/archives/C049HHMV9SM/p1691565944462639?thread_ts=1690907049.368789&cid=C049HHMV9SM) from the Slack thread. I don't think it's a feature request, as we've implemented a 30 second timer to prevent users requesting multiple codes at once, so I don't see why we wouldn't want this behaviour with Secondary Logins. Keen to hear if you agree!

@jliexpensify I think this is a feature request and ofcourse we could also think of this as a inconsitency bug. But Anyways I think its a valuable fix since we don't want users to request multiple codes and end up being blocked for sometime.  So It would be a go ahead from me. 
And I do have a proposal in mind to go with, before that do we want to confirm with product if we want to do this or should we go ahead with my review process?

**Comment by kaushiktd** [comment]
@jliexpensify I just checked after your comment. It is still reproducing with my current expensify code.

Also is that Slack link private? I am already a member of Expensify group on Slack and that link is showing error to me.

**Comment by jliexpensify** [comment]
Hi @kaushiktd - just confirming, you're referring to this [one](https://github.com/Expensify/App/issues/24312#issuecomment-1672525471)? If so, can you share a video? My app is playing up today, it's a bit laggy.

Also yes, that Slack link was posted to our internal team. Once I have some consensus, I'll update this GH!

**Comment by kaushiktd** [comment]
@jliexpensify I was referring to this one.


> Hmm I can't seem to repro it anymore. My guess is there was some sort of lag?

**Comment by jliexpensify** [comment]
@abdulrahuman5196 so this behaviour actually mirrors what we have over at www.expensify.com (i.e. when you're in your account, you can add a Secondary Login and request multiple magic codes without an error). The reasoning behind this is because once you're already in your account, the timer isn't needed - the timer is there at the login to stop bad actors spamming the Magic Code login.

So the team agreed that:

1. It's more likely a feature request
2. We should try and be consistent and require a timer when adding a Secondary Login

How do you think we should proceed? Open a Feature Request in #expensify-open-source, or work on this GH?

**Comment by jliexpensify** [comment]
@abdulrahuman5196 just bumping this - I guess we should go ahead and fix this, based off the internal thread? Do you have a proposal in mind?

**Comment by jliexpensify** [comment]
Bump @abdulrahuman5196 - did you have a fix in mind?

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
Based on this https://github.com/Expensify/App/issues/24312#issuecomment-1683361489

Since we treat this kind of a feature request. I am approving the first proposal with plan of action from @DanutGavrus here https://github.com/Expensify/App/issues/24312#issuecomment-1671195264

We should do this from the proposal

>move the timer's logic to a new reusable component and use it in both places.

🎀 👀 🎀
C+ Reviewed

**Comment by NikkiWines** [comment]
@francoisl is OOO right now but I'm happy to give this a once-over.

We did indeed intentionally add the 30s only to the sign-in view because that's a flow where we saw users get throttled regularly because they would get impatient and keep pressing the `Resend Magic Code` button 😅

We don't see nearly the same volume of throttling for users adding a secondary login / new contact method. It's not that you can't get throttled through that flow, just that it's a lower traffic flow that doesn't induce the same sort of urgency in users, for whatever reason. 

If we want to be consistent, we can add the 30s timer to the secondary login page, but I personally don't think it's really necessary unless we're seeing that users are regularly getting throttled through that flow.

**Comment by DanutGavrus** [comment]
@jliexpensify @srikarparsi Would you help us with some context from the internal discussion at https://expensify.slack.com/archives/C01SKUP7QR0/p1691974418915039? I don't have access to that. Should we proceed with https://github.com/Expensify/App/issues/24312#issuecomment-1690286314 or with https://github.com/Expensify/App/issues/24312#issuecomment-1690296363? Thank you!

**Comment by jliexpensify** [comment]
@DanutGavrus - so the internal discussion was partly in favour of consistency, and partly in favour of not needing a timer due to the reasons Nikki listed above. 

I favour consistency (I mentioned that as the reasoning for proceeding), but since @NikkiWines worked on implementing Magic Codes as one of the lead engineers, I think we should defer to you for the final decision. 

I went through some Secondary Login GH's and noticed that some users don't receive Magic Codes and need to log out/in again ([example](https://github.com/Expensify/Expensify/issues/290335)). Adding a timer wouldn't help with this issue right? If not, and users aren't really getting throttled with Secondary Logins, then maybe this isn't necessary.

**Comment by NikkiWines** [comment]
Yeah, @jliexpensify, for the issue you linked (and for the others that I've seen), a timer wouldn't resolve that error. We mainly saw issues before passwordless was out of a beta stage with users having sessions that used the old authentication methods (the issue you linked) or that they were using a phone number as a secondary login with an area code that's restricted. 

I chatted a bit with @johnmlee101 (the other lead eng. on this project) just to get a second opinion as well - seems like we're in agreement, and he can chime in if he has further thoughts. 

Overall, I get the consistency argument, but the timer was added to address the specific issue of users being unable to authenticate into their accounts because they were throttling themselves, and we're just not seeing that problem with secondary logins.
